### PR TITLE
Hold gestures

### PR DESCRIPTION
**Pull request by JoseExposito**
## Description:

As touchpad touches are generally fully abstracted, a client cannot currently know when a user is interacting with the touchpad without moving. This is solved by hold gestures.

Hold gestures are notifications about one or more fingers being held down on the touchpad without significant movement.

Hold gestures are primarily designed for two interactions:

 - Hold to interact: where a hold gesture is active for some time a menu could pop up, some object could be selected, etc.
 - Hold to cancel: where e.g. kinetic scrolling is currently active, the start of a hold gesture can be used to stop the scroll.

Unlike swipe and pinch, hold gestures, by definition, do not have movement, so there is no need for an "update" stage in the gesture.

## Status:

 - [libinput](https://gitlab.freedesktop.org/libinput/libinput/-/merge_requests/602): Merged and tagged as v1.18.900.
 - [Wayland protocol](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/81): As mentioned by @emersion : "The protocol itself looks fine to me. Would be nice to get feedback from Mutter/GTK maintainers, and a third implementation." Both Mutter and GTK are pending for review. This is the third implementation

## wlroots example:

To be able to test this feature, I modified `tinywl` so it sends gestures to clients. I didn't include the patch in this PR because it feels out of the scope of `tinywl`, but if you want to easily test, just apply this patch:

<details>
<summary>

0001-tinywl-add-gesture-support.patch (click to expand it)

</summary>

```diff
From 520ad9b383c1c4c57578e7a59a80db24c09eb6d4 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jos=C3=A9=20Exp=C3=B3sito?= <jose.exposito89@gmail.com>
Date: Mon, 12 Jul 2021 22:59:07 +0200
Subject: [PATCH] tinywl: add gesture support

---
 tinywl/tinywl.c | 131 ++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 131 insertions(+)

diff --git a/tinywl/tinywl.c b/tinywl/tinywl.c
index 2f2257ec..c6b66a3d 100644
--- a/tinywl/tinywl.c
+++ b/tinywl/tinywl.c
@@ -17,6 +17,7 @@
 #include <wlr/types/wlr_output.h>
 #include <wlr/types/wlr_output_layout.h>
 #include <wlr/types/wlr_pointer.h>
+#include <wlr/types/wlr_pointer_gestures_v1.h>
 #include <wlr/types/wlr_seat.h>
 #include <wlr/types/wlr_xcursor_manager.h>
 #include <wlr/types/wlr_xdg_shell.h>
@@ -46,8 +47,17 @@ struct tinywl_server {
 	struct wl_listener cursor_button;
 	struct wl_listener cursor_axis;
 	struct wl_listener cursor_frame;
+	struct wl_listener swipe_begin;
+	struct wl_listener swipe_update;
+	struct wl_listener swipe_end;
+	struct wl_listener pinch_begin;
+	struct wl_listener pinch_update;
+	struct wl_listener pinch_end;
+	struct wl_listener hold_begin;
+	struct wl_listener hold_end;
 
 	struct wlr_seat *seat;
+	struct wlr_pointer_gestures_v1 *gestures;
 	struct wl_listener new_input;
 	struct wl_listener request_cursor;
 	struct wl_listener request_set_selection;
@@ -530,6 +540,106 @@ static void server_cursor_frame(struct wl_listener *listener, void *data) {
 	wlr_seat_pointer_notify_frame(server->seat);
 }
 
+static void server_swipe_begin(struct wl_listener *listener, void *data) {
+	struct tinywl_server *server =
+		wl_container_of(listener, server, swipe_begin);
+	struct wlr_event_pointer_swipe_begin *event = data;
+
+	wlr_pointer_gestures_v1_send_swipe_begin(
+			server->gestures,
+			server->seat,
+			event->time_msec,
+			event->fingers);
+}
+
+static void server_swipe_update(struct wl_listener *listener, void *data) {
+	struct tinywl_server *server =
+		wl_container_of(listener, server, swipe_update);
+	struct wlr_event_pointer_swipe_update *event = data;
+
+	wlr_pointer_gestures_v1_send_swipe_update(
+			server->gestures,
+			server->seat,
+			event->time_msec,
+			event->dx,
+			event->dy);
+}
+
+static void server_swipe_end(struct wl_listener *listener, void *data) {
+		struct tinywl_server *server =
+		wl_container_of(listener, server, swipe_end);
+	struct wlr_event_pointer_swipe_end *event = data;
+
+	wlr_pointer_gestures_v1_send_swipe_end(
+			server->gestures,
+			server->seat,
+			event->time_msec,
+			event->cancelled);
+}
+
+static void server_pinch_begin(struct wl_listener *listener, void *data) {
+	struct tinywl_server *server =
+		wl_container_of(listener, server, pinch_begin);
+	struct wlr_event_pointer_pinch_begin *event = data;
+
+	wlr_pointer_gestures_v1_send_pinch_begin(
+			server->gestures,
+			server->seat,
+			event->time_msec,
+			event->fingers);
+}
+
+static void server_pinch_update(struct wl_listener *listener, void *data) {
+	struct tinywl_server *server =
+		wl_container_of(listener, server, pinch_update);
+	struct wlr_event_pointer_pinch_update *event = data;
+
+	wlr_pointer_gestures_v1_send_pinch_update(
+			server->gestures,
+			server->seat,
+			event->time_msec,
+			event->dx,
+			event->dy,
+			event->scale,
+			event->rotation);
+}
+
+static void server_pinch_end(struct wl_listener *listener, void *data) {
+	struct tinywl_server *server =
+		wl_container_of(listener, server, pinch_end);
+	struct wlr_event_pointer_pinch_end *event = data;
+
+	wlr_pointer_gestures_v1_send_pinch_end(
+			server->gestures,
+			server->seat,
+			event->time_msec,
+			event->cancelled);
+}
+
+static void server_hold_begin(struct wl_listener *listener, void *data) {
+	struct tinywl_server *server =
+		wl_container_of(listener, server, hold_begin);
+	struct wlr_event_pointer_hold_begin *event = data;
+
+	wlr_pointer_gestures_v1_send_hold_begin(
+			server->gestures,
+			server->seat,
+			event->time_msec,
+			event->fingers);
+}
+
+static void server_hold_end(struct wl_listener *listener, void *data) {
+	struct tinywl_server *server =
+		wl_container_of(listener, server, hold_end);
+	struct wlr_event_pointer_hold_end *event = data;
+
+	wlr_pointer_gestures_v1_send_hold_end(
+			server->gestures,
+			server->seat,
+			event->time_msec,
+			event->cancelled);
+}
+
 /* Used to move all of the data necessary to render a surface from the top-level
  * frame handler to the per-surface render function. */
 struct render_data {
@@ -918,6 +1028,25 @@ int main(int argc, char *argv[]) {
 	server.cursor_frame.notify = server_cursor_frame;
 	wl_signal_add(&server.cursor->events.frame, &server.cursor_frame);
 
+	server.swipe_begin.notify = server_swipe_begin;
+	wl_signal_add(&server.cursor->events.swipe_begin, &server.swipe_begin);
+	server.swipe_update.notify = server_swipe_update;
+	wl_signal_add(&server.cursor->events.swipe_update, &server.swipe_update);
+	server.swipe_end.notify = server_swipe_end;
+	wl_signal_add(&server.cursor->events.swipe_end, &server.swipe_end);
+
+	server.pinch_begin.notify = server_pinch_begin;
+	wl_signal_add(&server.cursor->events.pinch_begin, &server.pinch_begin);
+	server.pinch_update.notify = server_pinch_update;
+	wl_signal_add(&server.cursor->events.pinch_update, &server.pinch_update);
+	server.pinch_end.notify = server_pinch_end;
+	wl_signal_add(&server.cursor->events.pinch_end, &server.pinch_end);
+
+	server.hold_begin.notify = server_hold_begin;
+	wl_signal_add(&server.cursor->events.hold_begin, &server.hold_begin);
+	server.hold_end.notify = server_hold_end;
+	wl_signal_add(&server.cursor->events.hold_end, &server.hold_end);
+
 	/*
 	 * Configures a seat, which is a single "seat" at which a user sits and
 	 * operates the computer. This conceptually includes up to one keyboard,
@@ -935,6 +1064,8 @@ int main(int argc, char *argv[]) {
 	wl_signal_add(&server.seat->events.request_set_selection,
 			&server.request_set_selection);
 
+	server.gestures = wlr_pointer_gestures_v1_create(server.wl_display);
+
 	/* Add a Unix socket to the Wayland display. */
 	const char *socket = wl_display_add_socket_auto(server.wl_display);
 	if (!socket) {
-- 
2.31.1
```

</details>

Let me know if you are interested in the patch, I can create another PR.

## How to test it:

Clone and install the latest libinput:

```bash
$ git clone https://gitlab.freedesktop.org/libinput/libinput.git
$ cd libinput
$ meson build --prefix=/usr -Ddocumentation=false
$ ninja -C build && ninja -C build install
```

Update your `unstable/pointer-gestures/pointer-gestures-unstable-v1.xml` Wayland protocol:
https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/81

Reboot.

To test a real use case, clone my GTK branch:
https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/3454 (Also, this PR is waiting for code review, comments are welcome :smiley:)

Run my my patched version of `tinywl` in a TTY, `cd` into my GTK branch and:

```bash
$ meson build --prefix=`pwd`/build/install --buildtype=debug # Installing here to not replace your stable GTK
$ ninja -C build && \
   ninja -C build install && \
   GDK_BACKEND=wayland WAYLAND_DEBUG=1 XDG_DATA_DIRS=`pwd`/build/install/share:$XDG_DATA_DIRS LD_LIBRARY_PATH=`pwd`/build/install/lib64:$LD_LIBRARY_PATH ./build/install/bin/gtk4-demo
```

Scroll somewhere, for example in the code, and hold with 2 fingers to stop kinetic scrolling. The gestures demo should also work.

### CODE REVIEW

File: backend/wayland/seat.c
Comment (emersion) [inline backend/wayland/seat.c:722]: We need to check the version supported by the compositor here, and avoid sending this request if it's < 3. You can get the version with `zwp_pointer_gestures_v1_get_version`.

We also need to request v3 if supported by the compositor in `registry_global`.

We're also missing the cleanup logic, but all the other gesture types are missing it too, so arguably that's unrelated to this PR.

**Review by emersion — CHANGES_REQUESTED** [review]
All changes look good to me, except this comment about the Wayland backend.

**Comment by emersion** [comment]
Ah, and one more detail: it would be nice to guard the libinput switch changes behind `#ifdef LIBINPUT_EVENT_GESTURE_HOLD_BEGIN` and `#ifdef LIBINPUT_EVENT_GESTURE_HOLD_END`, to allow the build to succeed on older libinput.

For the wayland-protocol bits we'll have to bump the version requirement in `meson.build` once the changes are released upstream.

**Comment by emersion** [comment]
BTW, thanks for nicely splitting up the changes. Greatly appreciated, makes the review a lot smoother. :)

File: backend/wayland/seat.c
Comment (emersion) [reply backend/wayland/seat.c:722]: > We're also missing the cleanup logic, but all the other gesture types are missing it too, so arguably that's unrelated to this PR.

https://github.com/swaywm/wlroots/pull/3048

File: backend/wayland/backend.c
Comment (emersion) [inline backend/wayland/backend.c:228]: We should also accept v1 and v2, since v3 is new.

**Comment by JoseExposito** [comment]
Hi @emersion,

Thanks a lot for the quick code review and for your valuable input, I learned a couple of things I made wrong in the past :)

> We need to check the version supported by the compositor here, and avoid sending this request if it's < 3. You can get the version with zwp_pointer_gestures_v1_get_version.

Good catch, I squashed the change into the right commit:
https://github.com/swaywm/wlroots/pull/3047/commits/9b6decfac4424f0786c198cffee76684f5ec4f42#diff-e247db6c7cf2dc285345455e80ace7faf16f9532b64dd1dfbb8dbcca19932e61R738-R742

> We're also missing the cleanup logic, but all the other gesture types are missing it too, so arguably that's unrelated to this PR.

Thanks for fixing it on master. I rebased on the top of your commit and added the required clean up code:
https://github.com/swaywm/wlroots/pull/3047/commits/9b6decfac4424f0786c198cffee76684f5ec4f42#diff-e247db6c7cf2dc285345455e80ace7faf16f9532b64dd1dfbb8dbcca19932e61R507-R509

> We also need to request v3 if supported by the compositor in registry_global.
> We should also accept v1 and v2, since v3 is new.

Double mistake :man_facepalming: Sorry about that. I hope this fix is the good one:
https://github.com/swaywm/wlroots/pull/3047/commits/9b6decfac4424f0786c198cffee76684f5ec4f42#diff-42c3c70f6154a0cbcf9e6efe015d8b67029dd35ebd73ec34693c9cdf3489070eR228-R229


> Ah, and one more detail: it would be nice to guard the libinput switch changes behind #ifdef LIBINPUT_EVENT_GESTURE_HOLD_BEGIN and #ifdef LIBINPUT_EVENT_GESTURE_HOLD_END, to allow the build to succeed on older libinput.

I believe it is not possible to use `#ifdef` with enums, but I might be wrong (too many years writing JS).
However, we can check libinput's version in meson and use `add_project_arguments`, does that work? I pushed the change to its own patch: https://github.com/swaywm/wlroots/pull/3047/commits/9c99926b1de06c2f2e4775e653f343c2c4414f21


In order to keep the git log clean, I squashed my changes, so make sure to `git pull --rebase` in case you already cloned the branch.

Thanks again for your input!!

File: backend/wayland/backend.c
Comment (emersion) [inline backend/wayland/backend.c:229]: `zwp_pointer_gestures_v1_interface.version` is the version the header was generated with, not the version supported by the parent compositor. The `version` variable actually indicates the version supported by the parent compositor.

If possible, I'd prefer to avoid using `fmin`, I'm worried about implicit number conversion rules and rounding shenanigans. `version < 3 ? version : 3` would work.

**Comment by emersion** [comment]
> I believe it is not possible to use #ifdef with enums, but I might be wrong (too many years writing JS).

Ah, that's correct! The version check LGTM.

File: backend/wayland/backend.c
Comment (JoseExposito) [reply backend/wayland/backend.c]: Oooops I hope this time I got it right. Tanks for your patience!

**Comment by JoseExposito** [comment]
Thanks again for the code review @emersion. Unless you need me to fix or add something else, I think this is done until the Wayland protocol gets merged.

**Review by emersion — COMMENTED** [review]
Thanks, this iteration of the PR looks good to me!

**Review by emersion — APPROVED** [review]

**Review by bl4ckb0ne — APPROVED** [review]

**Comment by JoseExposito** [comment]
It looks like the protocol is getting the required ACKs, so, rebased on the top of master and updated this patch https://github.com/swaywm/wlroots/pull/3047/commits/b0187cc21754995b3794e8ac83f5afc6aecf9e75  so the check is performed in the appropriate meson file.

**Comment by JoseExposito** [comment]
[libinput 1.19](https://lists.freedesktop.org/archives/wayland-devel/2021-September/041977.html) and [wayland-protocols 1.23](https://lists.freedesktop.org/archives/wayland-devel/2021-September/041979.html) have been released with hold gestures.

I added an extra patch bumping the protocol. Let me know if I need to do anything else.

**Comment by bl4ckb0ne** [comment]
Now we have to wait for the package update in the CI. I'll do the alpine one later today.

**Comment by emersion** [comment]
Now waiting for FreeBSD: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=258538

**Comment by emersion** [comment]
Thanks!

**Pull request merged**
